File: multinode/reputation/reputation.go
// Copyright (C) 2021 Storj Labs, Inc.
// See LICENSE for copying information.

package reputation

import (
	"time"

	"storj.io/common/storj"
)

// AuditWindow contains audit count for particular time frame.
type AuditWindow struct {
	WindowStart time.Time `json:"windowStart"`
	TotalCount  int32     `json:"totalCount"`
	OnlineCount int32     `json:"onlineCount"`
}

// Audit contains audit reputation metrics.
type Audit struct {
	TotalCount      int64         `json:"totalCount"`
	SuccessCount    int64         `json:"successCount"`
	Alpha           float64       `json:"alpha"`
	Beta            float64       `json:"beta"`
	UnknownAlpha    float64       `json:"unknownAlpha"`
	UnknownBeta     float64       `json:"unknownBeta"`
	Score           float64       `json:"score"`
	SuspensionScore float64       `json:"suspensionScore"`
	History         []AuditWindow `json:"history"`
}

// Stats encapsulates node reputation data.
type Stats struct {
	NodeID               storj.NodeID `json:"nodeId"`
	NodeName             string       `json:"nodeName"`
	Audit                Audit        `json:"audit"`
	OnlineScore          float64      `json:"onlineScore"`
	DisqualifiedAt       *time.Time   `json:"disqualifiedAt"`
	SuspendedAt          *time.Time   `json:"suspendedAt"`
	OfflineSuspendedAt   *time.Time   `json:"offlineSuspendedAt"`
	OfflineUnderReviewAt *time.Time   `json:"offlineUnderReviewAt"`
	VettedAt             *time.Time   `json:"vettedAt"`
	UpdatedAt            time.Time    `json:"updatedAt"`
	JoinedAt             time.Time    `json:"joinedAt"`
}

File: web/multinode/src/api/reputation.ts
// Copyright (C) 2021 Storj Labs, Inc.
// See LICENSE for copying information.

import { APIClient } from '@/api/index';
import { Audit, AuditWindow, Stats } from '@/reputation';

/**
 * ReputationClient is a reputation api client.
 */
export class ReputationClient extends APIClient {
    private readonly ROOT_PATH: string = '/api/v0/reputation';

    /**
     * stats handles retrieval of a node reputation for particular satellite.
     * @param satelliteId - id of satellite.
     */
    public async stats(satelliteId: string): Promise<Stats[]> {
        const path = `${this.ROOT_PATH}/satellites/${satelliteId}`;

        const response = await this.http.get(path);

        if (!response.ok) {
            await this.handleError(response);
        }

        const result = await response.json();

        return result.map(
            (stats: Stats) => new Stats(
                stats.nodeId,
                stats.nodeName,
                new Audit(
                    stats.audit.totalCount,
                    stats.audit.successCount,
                    stats.audit.alpha,
                    stats.audit.beta,
                    stats.audit.unknownAlpha,
                    stats.audit.unknownBeta,
                    stats.audit.score,
                    stats.audit.suspensionScore,
                    stats.audit.history.map(
                        (auditWindow: AuditWindow) => new AuditWindow(
                            new Date(auditWindow.windowStart),
                            auditWindow.totalCount,
                            auditWindow.onlineCount,
                        ),
                    ),
                ),
                stats.onlineScore,
                new Date(stats.disqualifiedAt),
                new Date(stats.suspendedAt),
                new Date(stats.offlineSuspendedAt),
                new Date(stats.offlineUnderReviewAt),
                new Date(stats.updatedAt),
                new Date(stats.joinedAt),
            ),
        );
    }
}
